### ISSUE TITLE
[EPIC] Phase 2: SurVAE extension — surjections, stochastic transforms, examples

### ISSUE THREAD

**Issue by jejjohnson** [jejjohnson/gauss_flows#20]
Tracking issue for **Phase 2** of the gauss_flows refresh. Phase 1 ([#6](https://github.com/jejjohnson/gauss_flows/pull/6)) upgraded the repo scaffolding. Phase 2 builds the SurVAE extension: surjections, stochastic transforms, their composition factory, plus completion of the Phase-1 stubs, example notebooks called out in the [design docs](../jej_vc_snippets/design_docs/gauss_flows/), and a tranche of additions surfaced by reconciling against the [VincentStimper/normalizing-flows](https://github.com/VincentStimper/normalizing-flows) catalogue.

## Goal

Bring gauss_flows from "Gaussianization flows on top of FlowJax" to the full SurVAE hierarchy:

```
Transform (base)
├── Bijection            — FlowJax AbstractBijection (already shipped)
├── Surjection           — NEW: deterministic one way, stochastic the other
│   ├── InferenceSurjection  — exact likelihood
│   └── GenerativeSurjection — lower bound
└── StochasticTransform  — NEW: stochastic both ways, lower bound (e.g. VAE)
```

## Load-bearing architecture decisions

**Option A — parallel hierarchy** (confirmed):
- `AbstractSurjection(eqx.Module)` is its own base with `forward_and_log_det(x, key, cond=None)` and `inverse_and_log_det(z, key, cond=None)` — NOT a subclass of `flowjax.bijections.AbstractBijection`.
- `AbstractStochastic(AbstractSurjection)` layer for VAE-style transforms.
- New `SurVAEFlow` container threads PRNG keys explicitly through a mixed chain.
- Pure-bijection flows keep using `flowjax.Scan` / `Invert` / `Transformed` — zero regression.

**Conditional distribution interface** — structural `Protocol` in `src/gauss_flows/_src/_protocols.py`:

```python
@runtime_checkable
class ConditionalDistribution(Protocol):
    def sample(self, key, *, condition) -> Array: ...
    def log_prob(self, value, *, condition) -> Array: ...
```

Accepts numpyro `Distribution`, flowjax conditional wrappers, or custom classes.

### Wave 1 — can start immediately (after PR-0)

- [x] #7  — feat(pr-0): AbstractSurjection foundation + SurVAEFlow container
- [x] #8  — feat(pr-1): simple surjections (Abs, Sort, MaxPool2d, StochasticPermutation)
- [x] #9  — feat(pr-2): Slice + Augment surjections
- [x] #13 — feat(pr-6): finish DeepSigmoidCoupling + HistogramCDF stubs

### Wave 2 — after PR-2 merges

- [x] #10 — feat(pr-3): Uniform + Variational Dequantization
- [x] #11 — feat(pr-4): VAE stochastic transform

### Wave 3 — critical path close

- [ ] #12 — feat(pr-5): survae_flow + InverseFlow + ELBO training

## Roadmap — normflows-catalogue additions

Surfaced by the VincentStimper/normalizing-flows scrape. Independent of the SurVAE core; can land in any order.

### High-priority base distributions + transforms

- [x] #21 — feat(pr-9): GaussianMixture base distribution
- [x] #22 — feat(pr-10): GaussianPCA base distribution (low-rank + diag)
- [x] #23 — feat(pr-11): periodic + circular flow support (geophysical use case)
- [x] #24 — feat(pr-12): LULinearPermute bijection
- [ ] #25 — feat(pr-13): toy-target distribution zoo
- [x] #26 — feat(pr-14): BatchNorm bijection layer
- [ ] #27 — proposal(pr-15): ADR — multiscale flow pattern (Split/Merge vs Slice/Augment)
- [ ] #28 — feat(pr-16): conditional flow support — context threading

## Roadmap — flow_synthesizer-derived additions

Surfaced by the ACIDS-IRCAM flow_synthesizer scrape. The regression-via-flow patterns are the novel IP; a few supporting primitives come along for the ride.

### Regression + variational machinery

- [ ] #31 — feat(pr-17): ConditionalFlowRegressor recipe — p(y|x) via hypernet → flow params
- [ ] #32 — feat(pr-18): VariationalSigma — learnable Gaussian observation-noise posterior
- [ ] #33 — feat(pr-19): Multiplicative Normalizing Flows — Bayesian linear via flow-over-weights
- [ ] #34 — feat(pr-20): Flow-KL regression objective — closed-form Gaussian matching
- [ ] #36 — feat(pr-22): semi-supervised per-axis Gaussian-target objective

### Classic bijections (pedagogical / baselines)

- [ ] #35 — feat(pr-21): PlanarFlow and SylvesterFlow bijections

## Roadmap — docs PRs

- [ ] #14 — docs(pr-7a): existing-feature showcase notebooks (7 — can land any time)
- [ ] #15 — docs(pr-7b): surjection notebooks — toy_2d, slice_dim_reduction, augment_demo
- [ ] #16 — docs(pr-7c): VAE notebooks — vae_transform, hvae_stacked, mnist_vae
- [ ] #17 — docs(pr-7d): multi-scale image flow, numpyro_vae_svi, vae_flow_prior
- [ ] #18 — docs(pr-7e): PriorCVAE full multi-stage tutorial
- [ ] #29 — docs(pr-7f): base distributions tutorial (numpyro + gauss_flows extensions)

## Cleanup

- [ ] #19 — chore(pr-8): remove Phase-1 ty rule suppressions

## Backlog (not Phase 2)

- [ ] #30 — icebox: Phase 3 candidates — residual flows, HAIS, MH proposals, image-prior utilities

## Parallelism map

```
                SurVAE core
                ─────────────
PR-0  ──┬── PR-1 ──┐
        ├── PR-2 ──┼── PR-3 ──┐
        │          └── PR-4 ──┼── PR-5 ── PR-7b/c/d/e
        └── PR-6 ─────────────┘             │
                                            │
PR-7a (doc showcase, no gate) ──────────────┘

           normflows additions
           ───────────────────
PR-9, PR-10, PR-11, PR-12, PR-13, PR-14, PR-16 — all independent, land any time
PR-15 (multiscale ADR) blocked by PR-2
PR-7f (base-dists tutorial) blocked by PR-9 + PR-10

           flow_synthesizer additions
           ──────────────────────────
PR-17 (ConditionalFlowRegressor) blocked by PR-16 (conditional flows)
PR-18 (VariationalSigma) — independent
PR-19 (MNF Bayesian linear) — independent
PR-20 (FlowKL regression) blocked by PR-18
PR-21 (Planar + Sylvester) — independent
PR-22 (disentangling loss) — independent

PR-8 (ty cleanup) — incremental across all others
```

**Critical path**: PR-0 → PR-2 → PR-5. Everything else parallelises.

## Scope boundaries (what's NOT in Phase 2)

- `ArgmaxSurjection` (categorical flows, per Nielsen 2021) — out of scope for now; revisit if discrete-data support becomes a priority.
- Residual flows (iResNet / Residual Flow) — deferred to Phase 3; tracked in #30.
- MCMC-adjacent machinery (HAIS, MH proposals) — deferred to Phase 3; tracked in #30.
- `pyrox-gp` integration notebook — deferred until pyrox-gp lands in the ecosystem.

**Conditional flows** are **in scope** — see #28 (context threading through transforms and base dists).

## Reference implementations

- [didriknielsen/survae_flows](https://github.com/didriknielsen/survae_flows) — canonical PyTorch reference for SurVAE (Nielsen et al. NeurIPS 2020)
- [VincentStimper/normalizing-flows](https://github.com/VincentStimper/normalizing-flows) — the `normflows` PyPI package; source for the catalogue additions in issues #21–#28
- [acids-ircam/flow_synthesizer](https://github.com/acids-ircam/flow_synthesizer) — Esling et al. 2020; source for the regression / variational / MNF / disentangling additions in issues #31–#36
- [shayan-kousha/SurVAE](https://github.com/shayan-kousha/SurVAE) — older plain-JAX port (useful for sanity-checking MaxPool indexing, not for style)
- [design_docs/examples/*.md](../jej_vc_snippets/design_docs/gauss_flows/examples/) — per-class worked examples this plan is reconciled against

## Done when

- All code PRs merged and tests green
- All 17 core example notebooks + base-dists tutorial land as both `.py` source and executed `.ipynb`
- `[tool.ty.rules]` suppression block removed from `pyproject.toml`
- Release-please cuts a `0.2.0` (or `0.3.0`, depending on breakage) after PR-5 lands
